rust/src/bin/test-drive-net.rs

```diff
@@ -1,52 +1,72 @@
-    participant::Task,
 };
 
-/// Test-drive script of a (local, but networked) single-round federated
-/// learning session, intended for use as a mini integration test. It assumes
-/// that a [`Service`] is already running and listening to
-/// http://127.0.0.1:8081.
-///
-/// 10 [`Client`]s are spawned on the tokio event loop. This serves as a simple
-/// example of getting started with the project, and may later be the basis for
-/// more automated tests.
+#[derive(Debug, StructOpt)]
+#[structopt(name = "Test Drive")]
+struct Opt {
+    #[structopt(
+        default_value = "http://127.0.0.1:8081",
+        short,
+        help = "The URL of the coordinator"
+    )]
+    url: String,
+    #[structopt(default_value = "4", short, help = "The length of the model")]
+    len: u32,
+    #[structopt(
+        default_value = "1",
+        short,
+        help = "The time period at which to poll for service data, in seconds"
+    )]
+    period: u64,
+    #[structopt(default_value = "10", short, help = "The number of clients")]
+    nb_client: u32,
+}
+
+/// Test-drive script of a (local, but networked) federated
+/// learning session, intended for use as a mini integration test.
 #[tokio::main]
 async fn main() -> Result<(), ClientError> {
     let _fmt_subscriber = FmtSubscriber::builder()
         .with_env_filter(EnvFilter::from_default_env())
         .with_ansi(true)
         .init();
 
-    // dummy local model for clients
-    let model = Model::from_primitives(vec![0_f32, 1_f32, 0_f32, 1_f32].into_iter()).unwrap();
+    let opt = Opt::from_args();
 
-    let mut tasks = vec![];
-    for id in 0..10 {
-        let mut client = Client::new_with_addr(1, id, "http://127.0.0.1:8081")?;
+    // // dummy local model for clients
+    let len = opt.len as usize;
+    let model = Model::from_primitives(vec![0; len].into_iter()).unwrap();
+
+    let mut clients = FuturesUnordered::<JoinHandle<()>>::new();
+    for id in 0..opt.nb_client {
+        let mut client = Client::new_with_addr(opt.period, id, &opt.url)?;
         client.local_model = Some(model.clone());
-        let join_hdl = tokio::spawn(async move { client.during_round().await });
-        tasks.push(join_hdl);
+        let join_hdl = tokio::spawn(async move {
+            tokio::select! {
+                _ = signal::ctrl_c() => {}
+                result = client.start() => {
+                    error!("{:?}", result);
+                }
+            }
+        });
+        clients.push(join_hdl);
     }
-    println!("spawned 10 clients");
-
-    let mut summers = 0;
-    let mut updaters = 0;
-    let mut unselecteds = 0;
-    for task in tasks {
-        match task.await.or(Err(ClientError::GeneralErr))?? {
-            Task::Update => updaters += 1,
-            Task::Sum => summers += 1,
-            Task::None => unselecteds += 1,
+
+    // wait for all the clients to finish
+    loop {
+        if clients.next().await.is_none() {
+            break;
         }
     }
```

Comment:
this looks like some kind of black magic to me, maybe you can shed some light here. according to the docs `client.next()` creates a future which will result in the next element of the collection when awaiting it. up to this point none of the spawned tasks are executed and the first iteration of the loop calling `clients.next()` will start the execution of the first task.

let's assume the case where the first future containing a `join_hdl` of the collection doesn't terminate (because of no errors in the `client.start()` loop and no ctrl-c). why are the other futures executed? or equivalently, how is the `clients.next()` iterating over the collection and starting the execution of all tasks even though the docs say it only yields the next element when the future is run to completion?

Reply:
Good question!
I had to test it myself because the documentation is not really clear.. I created a small example to expain what happens. If it doesn't answer your question pls ask again. I will then explain it in more detail.

```rust
    // create a long running future
    let join_hdl = tokio::spawn(async move {
        tokio::time::delay_for(tokio::time::Duration::from_secs(2)).await;
        info!("Hello 0");
        tokio::time::delay_for(tokio::time::Duration::from_secs(2)).await;
        info!("Hello again 0");
    });
    clients.push(join_hdl);

    // create two futures that are already `ready`/resolved
    for id in 1..3 {
        let join_hdl = tokio::spawn(async move { info!("Hello {}", id) });
        clients.push(join_hdl);
    }
    // at this point none of the futures is polled/executed yet
    // they are just pushed into the FuturesUnordered set

    // By calling `next` on FuturesUnordered, the set starts to poll each future simultaneously.
    // It will poll as long as each future is resolved.
    // However, clients.next() will only await until one of the future has been resolved.
    // As described in the documentation ("A set of futures which may complete in any order.")
    // it does not have to be the first future that will be completed first.
    clients.next().await; // returns the result of Hello 1/2
    
    // prints Hello 1/2
    // prints Hello 1/2

    // The FuturesUnordered set will still poll the remaining futures.
   
    // prints Hello 0
    // prints Hello again 0

    // However, if we don't await them via `next` or with a delay, they will be dropped.
    tokio::time::delay_for(tokio::time::Duration::from_secs(10)).await;
```

Reply:
the missing piece of the puzzle was that `next()` pulls all futures of the set instead of just any future. thanks for the explanation!

Reply:
@finiteprods I updated my comment above with the answer to your question

Reply:
thanks @Robert-Steiner for the analysis! how about if a `Client` future hangs (rather than, return with an error) - is the observed behaviour still similar? I'm just wondering what we have to gain by using `FuturesUnordered`...
(but even if there isn't much to gain, I don't have any objections to using it)

Reply:
Yes the behaviour is still similar.
I agree. The only advantage of "FuturesUnordered" that I see is that (according to the documentation) this structure is optimized to manage a large number of futures. But I don't think that we can so many clients at once. I will change it back to `vec` because it makes the code easier to read.

Reply:
cool, thanks for this interesting thread :+1: